### PR TITLE
added encoders 3.2

### PR DESCRIPTION
**Pull request by Jolanrensen**
Duration, Period, ByteArray (Binary, now actually working)

added tests for: Instant datatype, duration, period, binary, and others
added datatypes: ByteArray (BinaryType), CalendarInterval, Nothing (All three not yet working)

metioned: https://github.com/JetBrains/kotlin-spark-api/issues/131

WIP

### CODE REVIEW

**Comment by Jolanrensen** [comment]
okay the Period encoder seems to be broken by Spark itself. Other than that, all 3.2 stuff seems to be working and the tests actually test the contents now instead of just showing it

**Comment by Jolanrensen** [comment]
Oh right and the KotlinReflection scala file is now more readable :)

File: kotlin-spark-api/3.2/src/test/kotlin/org/jetbrains/kotlinx/spark/api/ApiTest.kt [hunk 1]
```diff
@@ -318,24 +324,97 @@ class ApiTest : ShouldSpec({
...
+            should("handle Period Datasets") { // uses encoder
+                val periods = listOf(Period.ZERO, Period.ofDays(2))
+                val dataset = periods.toDS()
+
+                dataset.show(false)
+
+                dataset.collectAsList().let {
+                    it[0] shouldBe Period.ZERO
+
+                    // TODO this is also broken in Scala. It reports a Period of 0 instead of 2 days
```
Comment (asm0dey) [inline kotlin-spark-api/3.2/src/test/kotlin/org/jetbrains/kotlinx/spark/api/ApiTest.kt:369]: Please link relevant JIRA ticket

File: kotlin-spark-api/3.2/src/main/kotlin/org/jetbrains/kotlinx/spark/api/ApiV1.kt [hunk 2]
```diff
@@ -1192,7 +1201,7 @@ fun schema(type: KType, map: Map<String, KType> = mapOf()): DataType {
                     DoubleArray::class -> typeOf<Double>()
                     BooleanArray::class -> typeOf<Boolean>()
                     ShortArray::class -> typeOf<Short>()
-                    ByteArray::class -> typeOf<Byte>()
+//                    ByteArray::class -> typeOf<Byte>() handled by BinaryType
```
[the remark this answers is not in the document]
Comment (Jolanrensen) [reply kotlin-spark-api/3.2/src/main/kotlin/org/jetbrains/kotlinx/spark/api/ApiV1.kt:1204]: done

File: kotlin-spark-api/3.2/src/test/kotlin/org/jetbrains/kotlinx/spark/api/ApiTest.kt (hunk 1 above)
Comment (Jolanrensen) [reply kotlin-spark-api/3.2/src/test/kotlin/org/jetbrains/kotlinx/spark/api/ApiTest.kt]: https://issues.apache.org/jira/browse/SPARK-38317 Apparently it's expected behavior, strange

**Pull request merged**